### ISSUE TITLE
Result of inference is better with .eval() turned off

### ISSUE THREAD

**Issue by shen338** [layumi/Person_reID_baseline_pytorch#218]
First of all, thank you for this amazing work! 

I am using a ResNet model with batch norm in training. In inference, I have to run the network with different batch sizes (1~10). I tested the model in train and eval mode (basically manipulating the network with .eval() in pytorch). 

The result is surprising: 

Test in training mode (no .eval()) gives the best result. Results of testing in eval mode with .eval() is not as good as in training mode.  And if I feed the image one by one to the network, rather than batch processing, the model gives worst result. 

I wonder why this is the case. Inference in eval mode supposes to have better performance. I was training the model with batch size = 64, maybe the batch size difference between training and inference causes this problem? Can you think of any possibilities that can cause this problem? 

And another question, what's the common practice to run ReID models with smaller batch size, like 1~5, which is way smaller than the batch size in training. 

Thank you!

**Comment by layumi** [comment]
Hi @shen338 
Thank you. 
1. `eval()` disables dropout and affects batchnorm to calculate the online mean and std. If you set batchsize = 0, the feature will be 0, since the mean feature is itself. 
Generally, we use the eval to avoid the testing jittering.

2. Training with small batchsize is a trick to achieve better result. It is not always true.
